### PR TITLE
Support username and password in spark dependencies

### LINKED ISSUE #703: Support username and password in spark dependencies
So we should make it recognise secretName, so that the cluster doesn't get lots of failed spark-dependencies runs.

Also, would the maintainers consider doing a spark-dependencies run immediately upon installation (or five minutes after, possibly?) so that the operator (me in this case) knows something is wrong?

Please assign me to this issue, I would like to fix this.

### PR DESCRIPTION
**Pull request by haf-afa**
Fixes #703

Tests are failing on

´´´
verifying github.com/operator-framework/operator-sdk@v0.11.0: checksum mismatch
	downloaded: h1:If0MefBJjPGYAW6Os6azpTMoaFtsSlXRlQctF23/ORs=
	go.sum:     h1:tQumPT2UjD6uhggfAerRbPt+rWOPKC80DmgKUEqeGYo=
´´´

This is a bit strange since I don't get the same hash on my Windows (!) machine.

### CODE REVIEW

File: pkg/cronjob/spark_dependencies.go [hunk 1]
```diff
@@ -133,15 +134,27 @@ func getStorageEnvs(s v1.JaegerStorageSpec) []corev1.EnvVar {
...
 			{Name: "ES_CLIENT_NODE_ONLY", Value: strconv.FormatBool(s.Dependencies.ElasticsearchClientNodeOnly)},
 			{Name: "ES_NODES_WAN_ONLY", Value: strconv.FormatBool(s.Dependencies.ElasticsearchNodesWanOnly)},
 		}
+
+		// we add ES_USERNAME and ES_PASSWORD (as empty strings) if the secret is empty or if we know that we have both
+		// username and passwords in the options Map (if also there's a secret, it's up to the user to make sure it is
+		// right at runtime, as the secret can be used for more stuff than ES_USERNAME/ES_PASSWORD)
+		username, hasUsername := sFlagsMap["es.username"]
+		password, hasPassword := sFlagsMap["es.password"]
+		if len(s.SecretName) == 0 || (hasUsername && hasPassword) {
```
Comment (pavolloffay) [inline pkg/cronjob/spark_dependencies.go]: What if somebody is providing user/pass directly in the options?

Comment (haf-afa) [reply pkg/cronjob/spark_dependencies.go]: Like this?

https://github.com/jaegertracing/jaeger-operator/pull/705/files#diff-bc7e3f9ed38fb3c44b603fecfb113155R64-L67

**Comment by haf-afa** [comment]
```
  storage:
    cassandraCreateSchema: {}
    dependencies:
      enabled: true
      image: jaegertracing/spark-dependencies
      resources: {}
      schedule: 55 23 * * *
    elasticsearch:
      nodeCount: 3
      redundancyPolicy: SingleRedundancy
      resources:
        limits:
          memory: 16Gi
        requests:
          cpu: "1"
          memory: 16Gi
      storage: {}
    esIndexCleaner:
      enabled: true
      image: jaegertracing/jaeger-es-index-cleaner:1.14.0
      numberOfDays: 7
      resources: {}
      schedule: 55 23 * * *
    esRollover:
      image: jaegertracing/jaeger-es-rollover:1.14.0
      resources: {}
      schedule: '*/30 * * * *'
    options:
      es:
        server-urls: http://apm-es-http.monitoring.svc:9200
    secretName: jaeger
    type: elasticsearch
```

This is what it looks like after deploying one jaeger CRD instance.

This then is the index cleaner cronjob:

```yaml
containers:
          - args:
            - "7"
            - http://apm-es-http.monitoring.svc:9200
            envFrom:
            - secretRef:
                name: jaeger
            image: jaegertracing/jaeger-es-index-cleaner:1.14.0
```

but the spark deps:

```yaml
 containers:
          - env:
            - name: STORAGE
              value: elasticsearch
            - name: ES_NODES
              value: http://apm-es-http.monitoring.svc:9200
            - name: ES_CLIENT_NODE_ONLY
              value: "false"
            - name: ES_NODES_WAN_ONLY
              value: "false"
            image: jaegertracing/spark-dependencies
            imagePullPolicy: Always
            name: jaeger-spark-dependencies
            resources: {}
            terminationMessagePath: /dev/termination-log
            terminationMessagePolicy: File
          dnsPolicy: ClusterFirst
```

Missing the secret name. However; the code in spark_dependencies.go has the same stuff in it as the es_index_cleaner.go:


spark-code

I'm a bit confused now.

**Comment by haf-afa** [comment]
#651

**Comment by haf-afa** [comment]
So this whole PR is because the operator hasn't been released since the 4th September...

**Comment by jpkrohling** [comment]
@haf-afa, do you mean that this is fixed in master already, but not available in a released binary?

**Comment by haf-afa** [comment]
Yes

**Comment by haf-afa** [comment]
I'm not sure why badger failed; I don't think I've caused any problems there. This PR could be merged if you want some extra unit tests and a bit of a cleaning of the env vars (to avoid filling them with empty strings if there's a secret attached), but besides that, I'd expect the new release-to-be fixes my problem.

**Comment by pavolloffay** [comment]
> So this whole PR is because the operator hasn't been released since the 4th September... crycry

I think you could use `lastest` tag to pull in the latest changes of the operator.

**Comment by jpkrohling** [comment]
> I think you could use lastest tag to pull in the latest changes of the operator.

I think we are not publishing the images (again?).

**Comment by pavolloffay** [comment]
I have created https://github.com/jaegertracing/jaeger-operator/issues/715

**Comment by haf** [comment]
How do I set the version of the image in the operator?

**Comment by jpkrohling** [comment]
Just change the image in the operator.yaml:
https://github.com/jaegertracing/jaeger-operator/blob/b6211a465568a11dcfd3b3ecf7ff16b30d8a9915/deploy/operator.yaml#L19

**Comment by pavolloffay** [comment]
@haf-afa thanks for the contribution I will close this one per:

* dependencies cron job already consumes env vars from secret. It removes empty env vars from flags before adding them to the container spec.
